### DISCUSSION TITLE
Repository Statistics Built-In on Magit

### DISCUSSION THREAD

**Discussion by LionyxML** [magit/magit#4846] [Ideas]
Hello there @tarcius and friends!

I don't know if this has already been discussed or if already is implemented, but I could not find anything about it.  "Hear me out" :)

What do you guys think of a new feature that allows `magit` to give some statistics about the current repository.

I usually go back to "Fork" (https://git-fork.com/) only because I want to check some statistics. Mostly:

![image](https://user-images.githubusercontent.com/16169950/212724714-4134db06-e237-47dd-b5fc-c3414b3b12ae.png)
![image](https://user-images.githubusercontent.com/16169950/212724731-a61ae72f-7259-492c-9643-8acc6dae0ead.png)
![image](https://user-images.githubusercontent.com/16169950/212724769-027a399e-427f-45f4-a558-3c14cbc7f54e.png)
![image](https://user-images.githubusercontent.com/16169950/212724789-3ff44adc-ab59-4e05-8180-38efcd3a1478.png)
![image](https://user-images.githubusercontent.com/16169950/212724802-dd363b14-bfa2-4c5f-b1e9-01efe7cdf525.png)
![image](https://user-images.githubusercontent.com/16169950/212724941-c45a3593-61b6-4746-a204-487aaca22e7e.png)

I have no Idea how `Fork` does it. I guess it calculates it self rather depending on some sort of remote API, since I have several private repos not pushed to any remote that still presents these statistics.

Is it just me or anybody else would also like to see something along those lines on magit?

**Comment by tarsius** [comment]
> What do you guys think of a new feature that allows `magit` to give some statistics about the current repository.

I would like that feature too, but I don't want to implement it. I also don't think it should be part of `magit` itself. It could be a magit-specific package, which displays the statistics in a buffer whose mode derives from `magit-mode`, but it doesn't necessarily have to be implemented that way.

A command line tool, something integrated into Org, something that produces a webpage, and something that is integrated into Magit, would all be useful. And to get all of that, I think it would be best to begin with the command line tool and implement the others on top.

Personally I use a [fork](https://github.com/hoxu/gitstats) of the unmaintained `gitstats`. The result looks like [this](https://magit.vc/stats/magit/activity.html). It's okay, but I certainly wouldn't mind something prettier, with more statistics.

**Reply by LionyxML** [reply]
For some reason I did not use the "Reply feature" here but continued with new posts.

I just submitted a MELPA recipie for the .el package: https://github.com/melpa/melpa/pull/8391.

For a first version, does it look cool? :)
![image](https://user-images.githubusercontent.com/16169950/217373110-5940737a-c9a0-4d1e-8407-60c4c0cb1426.png)

**Comment by LionyxML** [comment]
Cool!

I think I might have an Idea where to start. For basic stats we can easily format git log to JSON.

With node (and adding this dependency is currently what makes me sad) I can sort of easily (not thinking about performance)  generate all statistics data in JSON format in a CLI tool.

A pre-built React App using ECharts or Highcharts might be served using this JSON as one of it's assets.

But all of this seems a bit "out of Emacs".

I am guessing doing it all in Elisp with built-in solutions for graphics, generating statistics (in not async mode), and so, would be another order of magnitude of effort.

**Comment by LionyxML** [comment]
Hello there @tarsius !

Just FYI. I took this little bull by the horns, lol :)

It is still very early in progress but it is in motion. I've been developing in my free time a `node executable` CLI tool for generating git report stats. The output can be a JSON (both to a file or stdout) or rendered HTML.

Anyone with the latest node installed to their system should have `npx` installed by default. So, to try this CLI tool you need to:
- cd into the git project root folder
- issue `npx magit-stats`

And that's it. Node downloads the tool and executes it. An HTML report is generated then.

All about it in this repo: https://github.com/LionyxML/magit-stats

I'd love some feedback :)

**Reply by tarsius** [reply]
(Mentioning a discussion unfortunately doesn't create a cross-reference, so I'll do that manually: https://github.com/melpa/melpa/pull/8391.)
